uast/transformer/ast.go

```diff
@@ -0,0 +1,189 @@
+		return Field{
+			Name:     uast.KeyRoles,
+			Op:       Var(vr),
+			Optional: vr + "_exists",
+		}
+	}
+	var rop ArrayOp
+	if len(roles) != 0 && op != nil {
+		rop = AppendRoles(op, roles...)
+	} else if op != nil {
+		rop = op
+	} else {
+		rop = Roles(roles...)
+	}
+	return Field{
+		Name: uast.KeyRoles,
+		Op: If(vr+"_exists",
+			Append(NotEmpty(Var(vr)), rop),
+			rop,
+		),
+	}
+}
+
+// ASTObjectLeft construct a native AST shape for a given type name.
+func ASTObjectLeft(typ string, ast ObjectOp) ObjectOp {
+	a := ast.Object()
+	if _, ok := a.GetField(uast.KeyRoles); ok {
+		panic("unexpected roles filed")
+	}
+	if typ != "" {
+		a.SetField(uast.KeyType, String(typ))
+	}
+	a.SetFieldObj(RolesField(typ + "_roles"))
+	return Part("_", a)
+}
+
+// ASTObjectRight constructs an annotated native AST shape with specific roles.
+func ASTObjectRight(typ string, norm ObjectOp, rop ArrayOp, roles ...role.Role) ObjectOp {
+	b := norm.Object()
+	if _, ok := b.GetField(uast.KeyRoles); ok {
+		panic("unexpected roles filed")
+	}
+	if typ != "" {
+		b.SetField(uast.KeyType, String(typ)) // TODO: "<lang>:" namespace
+	}
+	// it merges 3 slices:
+	// 1) roles saved from left side (if any)
+	// 2) static roles from arguments
+	// 3) roles from conditional operation
+	b.SetFieldObj(RolesFieldOp(typ+"_roles", rop, roles...))
+	return Part("_", b)
+}
+
+// ObjectRoles creates a shape that adds additional roles to an object.
+// Should only be used in other object fields, since it does not set any type constraints.
+func ObjectRoles(vr string, roles ...role.Role) Op {
+	return ObjectRolesCustom(vr, nil, roles...)
+}
+
+// ObjectRolesCustom is like ObjectRoles but allows to apecify additional conatraints for object.
+func ObjectRolesCustom(vr string, obj ObjectOp, roles ...role.Role) Op {
+	return ObjectRolesCustomOp(vr, obj, nil, roles...)
+}
+
+// ObjectRolesCustomOp is like ObjectRolesCustom but allows to apecify a custom roles lookup.
+func ObjectRolesCustomOp(vr string, obj ObjectOp, rop ArrayOp, roles ...role.Role) Op {
+	f := RolesFieldOp(vr+"_roles", rop, roles...)
+	if obj == nil {
+		obj = Fields{f}
+	} else {
+		o := obj.Object()
+		o.SetFieldObj(f)
+		obj = o
+	}
+	return Part(vr, obj)
+}
+
+// OptObjectRoles is like ObjectRoles, but marks an object as optional.
+func OptObjectRoles(vr string, roles ...role.Role) Op {
+	return Opt(vr+"_set", ObjectRoles(vr, roles...))
+}
+
+// Operator is a helper to make an AST node describing an operator.
+func Operator(vr string, lookup map[uast.Value]ArrayOp, roles ...role.Role) ObjectOp {
+	roles = append([]role.Role{
+		role.Expression, role.Operator,
+	}, roles...)
+	var opRoles Op = Roles(roles...)
+	if lookup != nil {
+		opRoles = AppendRoles(
+			LookupArrOpVar(vr, lookup),
+			roles...,
+		)
+	}
+	return Fields{
+		{Name: uast.KeyType, Op: String(uast.TypeOperator)},
+		{Name: uast.KeyToken, Op: Var(vr)},
+		{Name: uast.KeyRoles, Op: opRoles},
+	}
+}
+
+func uncomment(s string) (string, error) {
```

Comment:
This should be parametrized for the comment marks instead of hardcoding `//` and `/*`.

Reply:
It would make it more verbose, but it would be useful for `Javadoc`, I agree.

Are you thinking about this kind of signature?
```golang
func Uncomment(vr string, tokens [][2]string) Op
```

Reply:
That could work, yes.

Reply:
Found that it won't work - current version reverses the transform by checking for `\n`. Making it generic won't work this way. Still, we can provide one helper to remove a single token pair.